src/default-directory-searcher.coffee

```diff
@@ -0,0 +1,107 @@
+#
+# Implements thenable so it can be used with `Promise.all()`.
+class DirectorySearch
+  # Public: Creates a new DirectorySearch that will not start running until the
+  # `emitter` that is private to this file emits an event with the specified `id`.
+  constructor: (directory, delegate, options, id) ->
+    @task = new Task(require.resolve('./scan-handler'))
+    rootPaths = [directory.getPath()]
+    @promise = new Promise (resolve, reject) =>
+      @task.on('task:cancelled', reject)
+      emitter.once id, =>
+        @task.start(rootPaths, options.regexSource, options, resolve)
+    @task.on 'scan:result-found', delegate.didMatch
+    @task.on 'scan:file-error', delegate.didError
+    @task.on 'scan:paths-searched', delegate.didSearchPaths
+
+  # Public: Implementation of `then()` to satisfy the *thenable* contract.
+  # This makes it possible to use a `DirectorySearch` with `Promise.all()`.
+  #
+  # Returns `Promise`.
+  then: (args...) ->
+    @promise.then.apply(@promise, args)
+
+  # Public: Cancels the search.
+  cancel: ->
+    # This will cause @promise to reject.
+    @task.cancel()
+    null
+
+
+# Default provider for the `atom.directory-searcher` service.
+module.exports =
+class DefaultDirectorySearcher
+  # Public: Determines whether this object supports search for a `Directory`.
+  #
+  # * `directory` {Directory} whose search needs might be supported by this object.
+  #
+  # Returns a `boolean` indicating whether this object can search this `Directory`.
+  canSearchDirectory: (directory) -> true
+
+  # Public: Performs a text search for files in the specified `Directory`, subject to the
+  # specified parameters.
+  #
+  # Results are streamed back to the caller by invoking methods on the specified `delegate`.
+  #
+  # * `directory` {Directory} that has been accepted by this provider's `canSearchDirectory()`
+  # predicate.
+  # * `delegate` {Object} with the following properties:
+  #   * `didMatch` {Function} call with a search result structured as follows:
+  #     * `searchResult` {Object} with the following keys:
+  #       * `filePath` {String} absolute path to the matching file.
+  #       * `matches` {Array} with object elements with the following keys:
+  #         * `lineText` {String} The full text of the matching line (without a line terminator character).
+  #         * `lineTextOffset` {Number} (This always seems to be 0?)
+  #         * `matchText` {String} The text that matched the `regex` used for the search.
+  #         * `range` {Range} Identifies the matching region in the file. (Likely as an array of numeric arrays.)
+  #   * `didError` {Function} call with an Error if there is a problem during the search.
+  #   * `didSearchPaths` {Function} periodically call with the number of paths searched thus far.
+  # * `options` {Object} with the following properties:
+  #   * `regexSource` {String} regex to search with. Produced via `RegExp::source`.
```

Comment:
I think that `regexSource` should be a separate parameter, since it's so central to the point of the method. I would also say that the `delegate` and `options` objects could be merged. I don't think we should frame this like iOS/Cocoa, where the delegate is its own long-lived entity; it's just a collection of optional callbacks, which we're going to be creating for each call to `::search()`.

So I'd say a call should look like:

```coffee
search = searcher.search(theDirectory, "the-pattern", {
  ignoreCase: true,
  didError: ->
  didMatch: ->
  didSearchPaths: ->
})
```